File: src/tutorial/hello_object.cc
#include "tutorial/hello_object.hh"

#include "base/logging.hh"
#include "base/trace.hh"
#include "debug/Hello.hh"


using namespace std;

// constructor : pass the parameter object to SimObject
HelloObject::HelloObject(HelloObjectParams *params):
    SimObject(params),
    event([this]{processEvent();}, name()),
    goodbye(params->goodbye_object),
    // each params instantiation has a name
    // which comes from the python config file
    myName(params->name), 
    latency(params->time_to_wait),
    timesLeft(params->number_of_fires)
{
    DPRINTF(Hello, "[info] : created the Hello object\n");
    panic_if(!goodbye, "HelloObject must have a non-null GoodbyeObject\n");
}

void HelloObject::processEvent() {
    if (--timesLeft > 0) {
        DPRINTF(Hello, "[info] : processing Event : %d times left\n", timesLeft);
        schedule(event, curTick() + latency);
    } else {
        goodbye->sayGoodbye(myName);
        DPRINTF(Hello, "[info] : Done processing!!\n");
    }
}

void HelloObject::startup() {
    schedule(event, latency);
}



// implement for the parameter type
// implicitly created from the SimObject Python declaration, namely, the create function
// returns a new instantiation of the SimObject
HelloObject* HelloObjectParams::create() {
    return new HelloObject(this);
}

File: src/tutorial/HelloObject.py
from m5.params import *
from m5.SimObject import SimObject

class HelloObject(SimObject):
    # convention to have same name with class
    # type is the C++ class that you are wrapping with this Python SimObject
    type='HelloObject'

    # file that contains the declaration of the class used as the 'type' parameter
    # convention is to use lowercase and underscore with the same name
    cxx_header='tutorial/hello_object.hh'


    # adding parameters
    # they will be passed to the C++ object
    # Param.<TypeName>(default_value, "description")
    # Many other convenience parameters (e.g. Percent, Cycles, MemorySize)
    time_to_wait = Param.Latency("Time before firing the event")
    number_of_fires = Param.Int(1, "Number of times to fire the event before goodbye")

    goodbye_object = Param.GoodbyeObject("A goodbye object")


class GoodbyeObject(SimObject):
    type='GoodbyeObject'
    cxx_header='tutorial/goodbye_object.hh'

    bandwidth = Param.MemoryBandwidth('100MB/s', "BW to fill the buffer")
    buffer_size = Param.MemorySize('1kB', "size of the buffer to fill")

